### PR TITLE
subsys/nvs: Add option for sector size 64kB or larger

### PR DESCRIPTION
**Pull request by Marco-Peter**
NVS sectors are currently  limited to have a size below 64 kBytes
which makes the usage of nvs on some flash memories with large
erase sectors not possible.
This patch introduces an option to allow for sectors sizes
of 64kB or more.
Enabling this option results in two bytes larger ATE which
allows for a 4 bytes offset value.
For backwards compatibility, disabling this option does not
change the ATE format at all.
Tests are extended to check both configurations.

### CODE REVIEW

**Comment by Laczen** [comment]
Hi @Marco-Peter, thanks for creating the PR. I like the idea of adding support for larger sectors (for some of the stm deevices).

It seems to me that there is something wrong with the support of smaller sector sizes:
```
/* Allocation Table Entry */
struct nvs_ate {
#if CONFIG_NVS_LARGE_SECTORS
	/* Order to keep proper alignment within struct */
	u32_t offset;  /* data address within file system */
	u16_t len;     /* data length */
	u16_t id;      /* data id */
#else
	u16_t id;      /* data id */
	u16_t offset;  /* data address within file system */
	u16_t len;     /* data length */
#endif
	u8_t part;     /* part of a multipart data - future extension */
	u8_t crc8;     /* crc8 check of the entry */
} __packed;
```
For "small" sector sizes this limits the FS (not the sector) to being smaller than 64kB.

Maybe it would be better to use the part element to provide support for larger sectors: the offset in the sector becomes:
```
    off_in_sector = (part << 16) + offset
```
This would move the largest sector size to 8MB.

**Comment by Marco-Peter** [comment]
Hi @Laczen, thank you for your reply.
The field comments here are currently wrong. The function of the offset field is still the offset to the sector start as it was before.
I thought for a time to replace the offset with an absolute address and obviously forgot to revert the comment together with the field name when I discarded this idea.
I can push the correction this evening or I can wait for you to complete the review - whatever you prefer.

**Comment by Laczen** [comment]
@Marco-Peter, you can leave this correction for later. I am wondering, before the changes you could have a sectorsize of 3 pages, is this still supported ?

**Comment by Marco-Peter** [comment]
@Laczen This is indeed not possible. The problem is probably not the number of pages per sector, but the sector base address calculation, which fails when the sector size is not a power of 2.
If this is an issue, I can think of a possibility to resolve this by splitting the address into sector number and offset, whereas the offset data type could be 16 bit for the small sectors and 32 bit for the large sectors. In principle the same as before, just using two distinct variables.
In this case, should I also update the documentation? NVS_SECTOR_SIZE is still specified to be a power of two there.

**Comment by Laczen** [comment]
@Marco-Peter, nice work, thanks. I am currently stuck with how to keep a compatible solution with small sectors. I have not yet come up with a good solution.

**Comment by Marco-Peter** [comment]
@Laczen  Thanks and you are welcome :-)
Do you mean the compatibility on the data storage or something else? Since as long as the parameter is not willingly changed to a value over 16, the implementation should actually behave exactly the same as before. The only reason to change the parameter would be the need to have larger sectors, and a change of the sector size would "destroy" the storage anyway. Or am I missing something?

**Comment by Laczen** [comment]
> @Laczen Thanks and you are welcome :-)
> Do you mean the compatibility on the data storage or something else? Since as long as the parameter is not willingly changed to a value over 16, the implementation should actually behave exactly the same as before. The only reason to change the parameter would be the need to have larger sectors, and a change of the sector size would "destroy" the storage anyway. Or am I missing something?

The possible problem I see is that it is not known what the value of NVS_OFFSET_BIT is. Maybe it is possible to link this to the sector size. As long as the sector size is smaller then 32kB it is 16, when the sector size is bigger, NVS_OFFSET_BIT jumps to 24.

**Comment by Marco-Peter** [comment]
@Laczen The problem on using the sector size would be that this is not a compile time constant. Therefore it would not be possible to set the type of the sector offset depending on it, which is a requirement when it has to stay compatible with the current implementation.
One possibility I could imagine is to give a selection of maximum sector sizes in KConfig instead of the currently proposed NVS_OFFSET_BIT value.
Lets say, the user can select between 64, 256 and 1024 kB sector size limits with 64kB being the default. If larger sectors would be required in future, this list could easily be extended.

**Comment by Marco-Peter** [comment]
Rebased to the current master since quite a few things have changed over the last month (especially change to standard integers).
Keeping open the proposal from 11 May.

**Comment by Marco-Peter** [comment]
Rebased to insert commit 38f623d97cd97429689141aa460e0278f61c9bd6

**Comment by Marco-Peter** [comment]
Rebased and integrated commits 22be0d8fd5902da11c9bf55a44012abc7c2cf1c1 and 0171bd9578759acbf120c54790a575867848cf0c.
Applied some modifications on test_nvs_gc_corrupt_ate and test_nvs_gc_corrupt_close_ate to use proper alignment.

**Comment by Laczen** [comment]
Hi @Marco-Peter, sorry I didn't reply any sooner.

I am worried that these large sectors will make nvs very slow (the same holds for very large nvs filesystems). To validate that it has acceptable performance could you perform a test like:
```
int rc;
uint32_t i = 0;
while (fs->ate_wra > (fs->data_wra + 32)) {
  LOG_INF("Write id 0 iter %i", i);
  rc = nvs_write(&fs, 0, &i, sizeof(i));
  LOG_INF("Finished write id 0 [%d]", rc);
  while (uint32_t j = 0; j < i; j++) {
    rc = nvs_write(&fs, 1, j, sizeof(j));
  }
  i++;
}
```

BTW, this is not the worse case, the worse case is when you would write some data at the very start and do a rewrite at the very end of the sector.

File: subsys/fs/nvs/nvs.c [hunk 1]
```diff
@@ -17,6 +17,26 @@
...
 /* basic routines */
+
+/* Get the base address of the sector where the address points into */
+static inline uint32_t sector_base_addr(struct nvs_fs *fs, uint32_t addr)
+{
+	return fs->sector_size * (addr >> NVS_OFFSET_BITS);
+}
+
+/* Get the offset from the addresses to the next lower sector start */
+#if NVS_OFFSET_BITS > 16
```
Comment (sjg20) [inline subsys/fs/nvs/nvs.c:28]: Can you put the #if instead the function? What is the point of using uint16_t as a return value? May we wlll just use unsigned int?

File: subsys/fs/nvs/nvs.c [hunk 2]
```diff
@@ -298,7 +318,7 @@ static int nvs_ate_cmp_const(const struct nvs_ate *entry, uint8_t value)
 
 /* store an entry in flash */
 static int nvs_flash_wrt_entry(struct nvs_fs *fs, uint16_t id, const void *data,
-				size_t len)
+			       uint16_t len)
```
Comment (sjg20) [inline subsys/fs/nvs/nvs.c:321]: Why uint16_t here?

File: tests/subsys/fs/nvs/src/main.c [hunk 3]
```diff
@@ -72,13 +73,38 @@ void test_nvs_init(void)
...
+void test_nvs_init_large_sectors(void)
+{
+#if NVS_OFFSET_BITS > 16
+	init(65);
+#else
+	ztest_test_skip();
+#endif
+}
+
+/* nvs_al_size returns size aligned to fs->write_block_size */
```
Comment (sjg20) [inline tests/subsys/fs/nvs/src/main.c:97]: comment and function name don't match

File: subsys/fs/nvs/nvs_priv.h [hunk 4]
```diff
@@ -30,11 +41,18 @@ extern "C" {
...
-	uint16_t len;	/* data len within sector */
-	uint8_t part;	/* part of a multipart data - future extension */
-	uint8_t crc8;	/* crc8 check of the entry */
+#if NVS_OFFSET_BITS > 16
+	/* Order to keep proper alignment within struct */
+	uint32_t offset;   /* data offset within sector */
+	uint16_t len;      /* data length */
+	uint16_t id;       /* data id */
+#else
+	uint16_t id;       /* data id */
```
Comment (sjg20) [inline subsys/fs/nvs/nvs_priv.h:50]: Is this to optimise things for 16-bit microcontrollers or to save data space?

**Review by sjg20 — CHANGES_REQUESTED** [review]

**Comment by Marco-Peter** [comment]
@Laczen I did some tests on the qemu target writing a sequence of 32 IDs and changing values repeatedly over 3125 times which results in a total of 100'000 writes.

1k sectors:
Total required time: 133'760ms
Fastest time: 0ms
Longest time for one single nvs_write call: 50ms

32k sectors:
Total required time: 2'040'350ms
Fastest time for nvs_write: 0ms
Longest time for one single nvs_write call: 42'230ms

64k sectors:
Total required time: 3'930'620ms
Fastest time for nvs_write: 0ms
Longest time for one single nvs_write call: 168'470ms

128k sectors:
Total required time: 6'293'040ms
Fastest time for nvs_write: 0ms
Longest time for one single nvs_write call: 673'050ms

Although I would not take the timings for too real. The relation between them
is quite clear. Especially the sector change and erase takes a lot of time - even if it
is only a 32k sector.

Earlier you wrote that you would be interested in the support for those STM32 with large flash sectors. Is this still the case with those timings? Otherwise I probably will close the PR and propose a change so that also 64k are supported (changing sector_size in struct nvs_fs to be type size_t).

**Comment by Laczen** [comment]
> @Laczen I did some tests on the qemu target writing a sequence of 32 IDs and changing values repeatedly over 3125 times which results in a total of 100'000 writes.
> 
> 1k sectors:
> Total required time: 133'760ms
> Fastest time: 0ms
> Longest time for one single nvs_write call: 50ms
> 
> 32k sectors:
> Total required time: 2'040'350ms
> Fastest time for nvs_write: 0ms
> Longest time for one single nvs_write call: 42'230ms
> 
> 64k sectors:
> Total required time: 3'930'620ms
> Fastest time for nvs_write: 0ms
> Longest time for one single nvs_write call: 168'470ms
> 
> 128k sectors:
> Total required time: 6'293'040ms
> Fastest time for nvs_write: 0ms
> Longest time for one single nvs_write call: 673'050ms
> 
> Although I would not take the timings for too real. The relation between them
> is quite clear. Especially the sector change and erase takes a lot of time - even if it
> is only a 32k sector.
> 
> Earlier you wrote that you would be interested in the support for those STM32 with large flash sectors. Is this still the case with those timings? Otherwise I probably will close the PR and propose a change so that also 64k are supported (changing sector_size in struct nvs_fs to be type size_t).

Hi @Marco-Peter, thanks for doing the tests. I didn't think it would be so bad. The possibility of having to wait more than 10 minutes for a write IMHO means that it would be a nogo for 128kB sectors, even the worse write on 32kB sectors would probably trigger a reboot when a watchdog is used.

If you would anyhow like the PR to be merged (I will not block this) I would add a compile warning that "Use of large nvs sectors (> 16kB) is highly discouraged because write and/or read can take a (very) long time to complete".

For stm devices that have mixed sector size (4 times 16kB, 1 time 64kB and x times 128kB) the focus should be on using the 16kB sectors for nvs.

**Comment by Marco-Peter** [comment]
Hi @Laczen, I have run a few more tests, and it does not look that bad, when somebody is using larger data elements. To document this, I added a paragraph to the documentation.
As for the compiler warning, I would not know where to insert it, since the sector size usually is defined in runtime (the sample uses the flash information for this).
So I would like to merge the pull request.

File: subsys/fs/nvs/nvs_priv.h (hunk 4 above)
Comment (Marco-Peter) [reply subsys/fs/nvs/nvs_priv.h:48]: Primarily compatibility reasons. The ATEs are stored on the flash. Changing their format would brake an existing partition created with an older version.

File: subsys/fs/nvs/nvs.c (hunk 2 above)
Comment (Marco-Peter) [reply subsys/fs/nvs/nvs.c:321]: The offset is a 16 bit number. Directly taking it as uint16_t discloses the allowed range to the function caller and does not require any typecast in line 311/331.

**Comment by Marco-Peter** [comment]
Closing this PR, since write calls take a lot time to complete, even on 32kB sectors (see comment from 11 July). So working with 64kB or even larger sectors gets too impractical.

**Pull request closed without merging**